include/deal.II/matrix_free/fe_point_evaluation.h

```diff
@@ -1193,12 +1227,26 @@ FEPointEvaluation<n_components, dim, spacedim, Number>::integrate(
               }
 
           // compute
-          internal::integrate_add_tensor_product_value_and_gradient(
-            poly,
-            value,
-            gradient,
-            vectorized_points,
-            solution_renumbered_vectorized);
+          const unsigned int n_shapes = poly.size();
+          if (polynomials_are_hat_functions)
+            internal::integrate_add_tensor_product_value_and_gradient_linear(
```

Comment:
I don't mind to have the if-statement here. But what was the intention of this? `integrate_add_tensor_product_value_and_gradient_shapes()` (old) just calls this function.

Reply:
For `integrate()` this function did not exist, it is introduced in this PR. I think that `FEPointEvaluation` has to decide this now, since the shapes are computed in `reinit()` now, but not in the case of `polynomials_are_hat_functions`. Here we omit `values_of_array()` since it is done on the fly. Also, we only need the `vectorized_points` for the linear path in `evaluate()/integrate()` from now on, see https://github.com/dealii/dealii/pull/14982/commits/f1664887a27bb1d0386fb24ecef1bd8962756d1e .